aspnetcore/fundamentals/configuration/index.md

```diff
@@ -1,452 +1,1308 @@
+```json
+{
+  "section0": {
+    "key0": "value",
+    "key1": "value"
+  },
+  "section1": {
+    "key0": "value",
+    "key1": "value"
+  },
+  "section2": {
+    "subsection0" : {
+      "key0": "value",
+      "key1": "value"
+    },
+    "subsection1" : {
+      "key0": "value",
+      "key1": "value"
+    }
+  }
+}
+```
 
-### Switch mappings
+When the file is read into configuration, the following unique hierarchical keys are created to hold the configuration values:
 
-When manually building configuration with `ConfigurationBuilder`, a switch mappings dictionary can be added to the `AddCommandLine` method. Switch mappings allow key name replacement logic.
+* section0:key0
+* section0:key1
+* section1:key0
+* section1:key1
+* section2:subsection0:key0
+* section2:subsection0:key1
+* section2:subsection1:key0
+* section2:subsection1:key1
 
-When the switch mappings dictionary is used, the dictionary is checked for a key that matches the key provided by a command-line argument. If the command-line key is found in the dictionary, the dictionary value (the key replacement) is passed back to set the configuration. A switch mapping is required for any command-line key prefixed with a single dash (`-`).
+### GetSection
 
-Switch mappings dictionary key rules:
+[IConfiguration.GetSection](/dotnet/api/microsoft.extensions.configuration.iconfiguration.getsection) extracts a configuration subsection with the specified subsection key.
 
-* Switches must start with a dash (`-`) or double-dash (`--`).
-* The switch mappings dictionary must not contain duplicate keys.
+To return an [IConfigurationSection](/dotnet/api/microsoft.extensions.configuration.iconfigurationsection) containing only the key-value pairs in `section1`, call `GetSection` and supply the section name:
 
-In the following example, the `GetSwitchMappings` method allows command-line arguments to use a single dash (`-`) key prefix and avoid leading subkey prefixes.
+```csharp
+var configSection = _config.GetSection("section1");
+```
 
-[!code-csharp[](index/sample/CommandLine/Program.cs?highlight=10-19,32)]
+Similarly, to obtain the values for keys in `section2:subsection0`, call `GetSection` and supply the section path:
 
-Without providing command-line arguments, the dictionary provided to `AddInMemoryCollection` sets the configuration values. Run the app with the following command:
+```csharp
+var configSection = _config.GetSection("section2:subsection0");
+```
 
-```console
-dotnet run
+`GetSection` never returns `null`. If a matching section isn't found, an empty `IConfigurationSection` is returned.
+
+### GetChildren
+
+A call to [IConfiguration.GetChildren](/dotnet/api/microsoft.extensions.configuration.iconfiguration.getchildren) on `section2` obtains an `IEnumerable<IConfigurationSection>` that includes:
+
+* `subsection0`
+* `subsection1`
+
+```csharp
+var configSection = _config.GetSection("section2");
+
+var children = configSection.GetChildren();
 ```
 
-The cons
```

Comment:
This doesn't match the naming scheme from the other providers. AddDatabase?

Reply:
This example is missing a few discailmers.
1) This reads the whole table into config up front, it does not query the database on a per key basis.
2) Reload on change is not implemented
3) You would need some way to pass a connection string into this provider, likely from a secondary config builder.

Reply:
> This doesn't match the naming scheme from the other providers. AddDatabase?

I renamed it from `AddEntityFrameworkConfig` (Steve's original name) to `AddEFConfiguration` ... but they both seem to generally match. Isn't that ok?

> You would need some way to pass a connection string into this provider, likely from a secondary config builder.

That's what Steve had there ...

```csharp
.AddEFConfiguration(options =>
    options.UseSqlServer(connectionStringConfig.GetConnectionString("DefaultConnection"))
```

We're not generally documenting anything other than the in-memory dB. We want the demos based on EF to run without issues everywhere, so that's why most of the samps go with the in-memory dB.

Reply:
You don't need to wire up Sql in the sample, but it's important to mention that you may need a secondary config builder to get the connection strings.